### PR TITLE
programmatically create string content

### PR DESCRIPTION
**Pull request by Maoni0**
## Summary

We should be discouraging folks to use fixed to change the string content and recommend to use string.Create if they want to programmatically build a string.

@gewarren, do we have a code sample somewhere that uses string.Create? The doc for it seems quite brief. 

cc @jkotas, per our conversation.

### CODE REVIEW

**Comment by jkotas** [comment]
The samples are in dotnet/samples repo. E.g. the sample this is talking about is at https://github.com/dotnet/samples/blob/master/snippets/csharp/how-to/strings/ModifyStrings.cs.

I would delete the bad example from it. The samples should be 100% production quality code that people can copy. We should not have any broken buggy code in samples to demonstrate what a bad code look like.

**Comment by jkotas** [comment]
I do not see any sample for string.Create. I think it would be fine to add it to the ModifyStrings.cs sample above.

File: docs/csharp/how-to/modify-string-contents.md
Comment (gewarren) [inline docs/csharp/how-to/modify-string-contents.md]:
```suggestion
## Programmatically build up string content
```

File: docs/csharp/how-to/modify-string-contents.md
Comment (gewarren) [inline docs/csharp/how-to/modify-string-contents.md]:
```suggestion
Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET core added a <xref:System.String.Create> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations. 
```

File: docs/csharp/how-to/modify-string-contents.md
Comment (gewarren) [inline docs/csharp/how-to/modify-string-contents.md]:
```suggestion
After a string is created, you could use **unsafe** code to modify its content, but this is strongly discouraged. Unsafe code bypasses many of the features of .NET designed to minimize certain types of bugs in code and may prevent using some features that only work with normal code. The following example is provided for those rare situations where you want to modify a string in-place using unsafe code. The example shows how to use the `fixed` keyword. The `fixed` keyword prevents the garbage collector from moving the string object in memory while code accesses the memory using the unsafe pointer. It also demonstrates one possible side effect of unsafe operations on strings that results from the way that the C# compiler stores (interns) strings internally. In general, you shouldn't use this technique unless it is absolutely necessary. You can learn more in the articles on [unsafe](../language-reference/keywords/unsafe.md) and [fixed](../language-reference/keywords/fixed-statement.md). The API reference for <xref:System.String.Intern%2A> includes information on string interning.
```

Comment (gewarren) [reply docs/csharp/how-to/modify-string-contents.md]: I removed the GC acronym since it's not mentioned after this at all, so it's unnecessary.

**Review by gewarren — COMMENTED** [review]
I'll let @BillWagner comment on where the new String.Create code should live. We're in the midst of some changes for where we're keeping code snippets.

**Comment by Maoni0** [comment]
here's an example of string.Create if you need it -

```csharp
// constructing a string from a char array, prefix it with some additional characters
char[] chars = { 'a', 'b', 'c', 'd', '\0' };
int length = chars.Length + 2;        
string result = string.Create(length, chars, (Span<char> strContent, char[] charArray) =>
{
    strContent[0] = '0';
    strContent[1] = '1';
    for (int i = 0; i < charArray.Length; i++)
    {
        strContent[i + 2] = charArray[i];
    }
});
```

**Comment by BillWagner** [comment]
@gewarren 

> ... where the new String.Create code should live. We're in the midst of some changes for where we're keeping code snippets.

We're still finalizing some concerns on that. For now, let's put the new sample with the other samples for this article.  We won't have the plan in place quickly enough, and this is an important change to publish.

**Comment by Maoni0** [comment]
I thought about this some more and mostly just deleted the paragraph that talks about using fixed to modify the string content. and I will submit a PR in the samples repo to replace the fixed sample with the str.Create sample. it's unnecessary to talk so much about fixed and unsafe code in general when we are strongly discouraging folks from using it. also the original content didn't really emphasize the fact it's modifying the string content in the fixed block - it's fine to use the fixed block if you are *only* reading from the string content.

please take another look - @jkotas, @gewarren, @BillWagner

**Comment by Maoni0** [comment]
https://github.com/dotnet/samples/pull/1960

**Review by jkotas — APPROVED** [review]

File: docs/csharp/how-to/modify-string-contents.md
Comment (pkulikov) [inline docs/csharp/how-to/modify-string-contents.md]:
```suggestion
Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET Core added a <xref:System.String.Create%2A?displayProperty=nameWithType> method that allows you to programmatically fill in the character content of a string via a callback while avoiding the intermediate temporary string allocations.
```

File: docs/csharp/how-to/modify-string-contents.md
Comment (pkulikov) [inline docs/csharp/how-to/modify-string-contents.md]: There is unrelated paragraph between the example and the paragraph that describes the example. Should the example be moved one paragraph up?

Comment (Maoni0) [reply docs/csharp/how-to/modify-string-contents.md]: sounds good; I'll move it up

**Review by BillWagner — COMMENTED** [review]
Commenting instead of reviewing so it stays in my queue.

This is great @Maoni0 Once we've merged the corresponding sample, I'll :shipit:

**Comment by Maoni0** [comment]
I've addressed all feedback in the latest commit.

**Review by BillWagner — APPROVED** [review]
This LGTM.

@gewarren Did you have any final comments?

I've merged the samples PR, so this is ready if you're ready to approve it.

**Comment by BillWagner** [comment]
closing and reopening for a fresh build.

**Comment by BillWagner** [comment]
I checked the review site, and the samples are included correctly.

File: docs/csharp/how-to/modify-string-contents.md
Comment (gewarren) [inline docs/csharp/how-to/modify-string-contents.md]:
```suggestion
Since strings are immutable, the previous examples all create temporary strings or character arrays. In high performance scenarios, it may be desirable to avoid these heap allocations. .NET Core provides a <xref:System.String.Create%2A?displayProperty=nameWithType> method that allows you to programmatically fill in the character content of a string via a callback while avoiding intermediate temporary string allocations.
```

**Review by gewarren — APPROVED** [review]
Approved with one small suggestion. .NET Core "added" makes me ask the question "when?", so I just changed it to "includes".

**Comment by Maoni0** [comment]
incorporated latest feedback

**Pull request merged**
